selfie.c

```diff
@@ -7765,6 +7888,379 @@ char* smt_binary(char* opt, char* op1, char* op2) {
   return string;
 }
 
+char* smt_ternary(char* opt, char* op1, char* op2, char* op3) {
+  char* string;
+
+  string = string_alloc(1 + string_length(opt) + 1 + string_length(op1) + 1 + string_length(op2) + 1 + string_length(op3) + 1);
+
+  sprintf4(string, "(%s %s %s %s)", opt, op1, op2, op3);
+
+  return string;
+}
+
+uint64_t find_merge_location(uint64_t beq_imm) {
+  uint64_t original_pc;
+  uint64_t original_imm;
+  uint64_t merge_location;
+
+  // assert: the current instruction is a symbolic beq instruction
+  original_pc = pc;
+  original_imm = imm;
+
+  // examine last instruction before target location of the beq instruction
+  pc = pc + (beq_imm - INSTRUCTIONSIZE);
+
+  // we need to know which instruction it is
+  fetch();
+  decode();
+  
+  if (is != JAL)
+    // no jal instruction -> end of if without else branch
+    // merge is directly at target location of the beq instruction possible
+    merge_location = original_pc + beq_imm;
+  else {
+    if (signed_less_than(imm, 0) == 0) { 
+      // jal with positive imm -> end of if with else branch
+      // we have to skip the else branch in order to merge afterwards
+      merge_location = pc + imm;
+
+      pc = original_pc + INSTRUCTIONSIZE;
+    } else
+      // jal with negative imm -> end of loop body
+      // merge is only outside of the loop possible
+      merge_location = pc + INSTRUCTIONSIZE;
+  }
+
+  // we need to check if we are inside of a recursion before we reach the merge location
+  // if we are, we merge only when the recursion is finished
+  while (pc != merge_location) {
+    fetch();
+    decode();
+
+    if (is == JAL)
+      if (is_start_of_procedure_prologue(pc + imm)) {
```

Comment:
Same here. Check if JAL uses RA register.

Reply:
Actually the name of the procedure is a bit misleading, it is not only enough to check if we call a procedure, we also have to check that we are currently in this procedure which we call. I have improved that.